Disallow var syntax in string interpolation

c42f opened the pull request:
The `var"##"` syntax should be disabled in string interpolation.

Also work around a subtle problem with command interpolation's use of Meta.parse in non-greedy mode, combined with the new `var` syntax which would cause `Base.shell_parse("\"(\$var)\"")` to miss the trailing `)` character.

This is a quick fix for a bit of fallout from #32408 which affects PyCall's build script, among other things.

Fixes #32951

c42f committed:
Disallow var syntax in string interpolation

The var"##" syntax should be disabled in string interpolation.

JeffBezanson commented on src/julia-parser.scm:
```diff
@@ -2285,10 +2285,14 @@
-           (if (and (eq? t 'var) (eqv? (peek-token s) #\") (not (ts:space? s)))
+           (if (and (eq? t 'var)
+                    (if (not (or (ts:pbtok s) (ts:last-tok s)))
+                      ;; Hack to preserve (io.pos (ts:port s)) for non-greedy Meta.parse
```
Indentation is off here

c42f replied:
Oops thanks. I've adjusted my autoindent settings.

JeffBezanson commented on src/julia-parser.scm:
```diff
@@ -2285,10 +2285,14 @@
+                    (if (not (or (ts:pbtok s) (ts:last-tok s)))
+                      ;; Hack to preserve (io.pos (ts:port s)) for non-greedy Meta.parse
+                      (eqv? (peek-char (ts:port s)) #\"))
+                    (and (eqv? (peek-token s) #\") (not (ts:space? s))))
```
I think it would be clearer to nest this inside the previous `if` statement.

c42f replied:
This ordering was intentional: we must run `peek-token` before the `take-token` further down.  However, we musn't run `peek-token` unless necessary.

It's delicate I'm afraid.

Perhaps I'll just add an extra `peek-token` down below and rearrange as suggested. That would be clearer if (very slightly) redundant.

c42f committed:
Disallow `var` syntax in command interpolations

This is special cased for compatibility. A more general fix would be to
make cmd interpolation syntax exactly the same as string interpolation.

c42f commented:
Right I've sorted those problems out!

I also added a workaround and test to preserve the behavior of `Base.shell_parse` (and hence `@cmd` interpolation) with `$var` interpolation. This is ugly but the alternative is a non backward compatible change to the way that `@cmd` handles `$` interpolation in general.

I think we *should* change the way `$` interpolation works in `cmd`, but I don't want to couple that breaking change with getting this bug fixed.

c42f commented:
CI errors look unrelated (win64 — temp files; freebsd — an internal build slave error).

c42f commented:
@JeffBezanson can you live with the ugly `shell_parse` compatibility workaround I've got here? If so I'd like to merge this and come back to the inconsistency between cmd interpolation vs string interpolation another time.

MikeInnes commented:
This bug breaks both Juno and the CUDA stack, so it'd be great to get the fix in quickly.

chriscoey commented:
bump
